manual/src/refman/extensions/tail_mod_cons.etex

```diff
@@ -0,0 +1,548 @@
+Please either mark the called function with
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+Here the "append_flatten" helper is not annotated with
+"[\@tail_mod_cons]", so the calls "append_flatten xs xss" and "flatten
+xss" will \emph{not} be tail calls. The right fix here is to mark
+"append_flatten" to be tail-mod-cons.
+
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+    let[@tail_mod_cons] rec append_flatten xs xss =
+      match xs with
+      | [] -> flatten xss
+      | x :: xs -> x :: append_flatten xs xss
+    in append_flatten xs xss
+\end{verbatim}
+
+The same warning occurs when "append_flatten" is a non-tail-mod-cons
+function of the same recursive group; using the tail-mod-cons
+transformation is a property of individual functions, not whole
+recursive groups.
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+  append_flatten xs xss
+
+and append_flatten xs xss =
+  match xs with
+  | [] -> flatten xss
+  | x :: xs -> x :: append_flatten xs xss
+
+| append_flatten xs xss
+  ^^^^^^^^^^^^^^^^^^^^^
+Warning 72 [tmc-breaks-tailcall]: This call is in tail-modulo-cons position in a TMC function,
+but the function called is not itself specialized for TMC,
+so the call will not be transformed into a tail call.
+Please either mark the called function with
+the [@tail_mod_cons] attribute, or mark this call with
+the [@tailcall false] attribute to make its non-tailness explicit.
+\end{verbatim}
+
+Again the fix is to specialized "append_flatten" as well:
+\begin{verbatim}
+let[@tail_mod_cons] rec flatten = function
+| [] -> []
+| xs :: xss ->
+  append_flatten xs xss
+
+and[@tail_mod_cons] append_flatten xs xss =
+  match xs with
+  | [] -> flatten xss
+  | x :: xs -> x :: append_flatten xs xss
+\end{verbatim}
+
+Non-recursive functions can also be marked "[\@tail_mod_cons]"; this is
+typically useful for local bindings to recursive functions.
+
+Incorrect version:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+                           ^^^^^^^^^^^^^^^^^^^^^
+Warning 51 [wrong-tailcall-expectation]: expected tailcal
+\end{verbatim}
+
+Recommend fix:
+\begin{verbatim}
+let[@tail_mod_cons] rec map_vars f exp =
+  let[@tail_mod_cons] self exp = map_vars f exp in
+  match exp with
+  | Var v -> Var (f v)
+  | Let ((v, def), body) ->
+    Let ((f v, self def), (self[@tailcall]) body)
+\end{verbatim}
+
+In other cases, there is no benefit in making the called function
+tail-mod-cons, or it is not possible, for example it is a function
+parameter (the transformation only works with calls to
+known functions).
+
+For example, consider a substitution function on binary trees:
```

Comment:
"For example" was couple lines before. May use "To illustrate" instead?

Reply:
I'm told that English readers are not averse to repetition, so I mostly go for consistent usage in technical prose.

Reply:
Most English readers in our field are not native-English speakers. ;-)

Reply:
It's not a question of English vs. other languages; it's a question of technical writing vs. other writing styles.  In technical writing, clarity and precision matter more than elegance.  Using several different words to designate the same thing may look elegant, but it can also be confusing.  So, in technical writing, pick one term for a given concept and use it as many times as necessary.

Reply:
I am not arguing. It is just a funny thing to point it out.